docs/proposal/20240308-vpc-infra-creation.md

```diff
@@ -0,0 +1,393 @@
+	VPC *VPCResourceReference `json:"vpc,omitempty"`
+}
+
+// IBMVPCClusterStatus defines the observed state of IBMVPCCluster.
+type IBMVPCClusterStatus struct {
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// Conditions defines current service state of the load balancer.
+	// +optional
+	Conditions capiv1beta1.Conditions `json:"conditions,omitempty"`
+
+	// ControlPlaneLoadBalancerState is the status of the load balancer.
+	// Deprecated: rely on NetworkStatus instead.
+	// +optional
+	ControlPlaneLoadBalancerState *VPCLoadBalancerState `json:"controlPlaneLoadBalancerState,omitempty"`
+
+	// COSInstance is the reference to the IBM Cloud COS Instance used for the cluster.
+	COSInstance *VPCResourceReference
+
+	// networkStatus is the status of the VPC network in its entirety resources.
+	NetworkStatus *VPCNetworkStatus `json:"networkStatus,omitempty"`
+
+	// ready is true when the provider resource is ready.
+	// +kubebuilder:default=false
+	Ready bool `json:"ready"`
+
+	// resourceGroup is the reference to the IBM Cloud VPC resource group under which the resources will be created.
+	ResourceGroup *ResourceReference `json:"resourceGroupID,omitempty"`
+
+	// Deprecated: rely on NetworkStatus instead.
+	Subnet Subnet `json:"subnet,omitempty"`
+
+	// Deprecated: rely on NetworkStatus instead.
+	VPC VPC `json:"vpc,omitempty"`
+
+	// Deprecated: rely on ControlPlaneEndpoint
+	VPCEndpoint VPCEndpoint `json:"vpcEndpoint,omitempty"`
+}
+
+// VPCNetworkStatus provides details on the status of VPC network resources.
+type VPCNetworkStatus struct {
+	// computeSubnets references the VPC Subnets for the cluster's Data Plane.
+	// +optional
+	ComputeSubnets []*VPCResourceReference `json:"computeSubnets,omitempty"`
+
+	// controlPlaneSubnets references the VPC Subnets for the cluster's Control Plane.
+	// +optional
+	ControlPlaneSubnets []*VPCResourceReference `json:"controlPlaneSubnets,omitempty"`
+
+	// loadBalancers references the VPC Load Balancer's for the cluster.
+	// +optional
+	LoadBalancers []VPCLoadBalancerStatus `json:"loadBalancers,omitempty"`
+
+	// publicGateways references the VPC Public Gateways for the cluster.
+	// +optional
+	PublicGateways []*VPCResourceReference `json:"publicGateways,omitempty"`
+
+	// securityGroups references the VPC Security Groups for the cluster.
+	// +optional
+	SecurityGroups []*VPCResourceReference `json:"securityGroups,omitempty"`
+
+	// vpc references the IBM Cloud VPC.
+	// +optional
+	VPC *VPCResourceReference `json:"vpc,omitempty"`
+}
+
+// VPCResourceReference identifies a resource by id and type.
+type VPCResourceReference struct {
+	// controllerCreated indicates whether the resource is created by the CAPI controller.
+	ControllerCreated bool `json:"controllerCreated,omitempty"`
+
+	// crn defines the IBM Cloud CRN of the resource.
+	// +optional
+	CRN *string `json"crn,omitempty"`
+
+	// id defines the ID of the IBM Cloud resource.
+	ID string `json:id"`
```

Comment:
is both crn and id required, doen't crn contains id?

Reply:
CRN is another nice to have, as it includes the account Id, region, etc.

I can try to determine whether including the Id at all is useful (Id was the default required field), to potentially drop it in this definition. I just don't have enough information ATM.

Reply:
Dropped ID